### PR TITLE
Issue 24519: Created method lazilyAsync for both DSLs and adds section in docs

### PR DESCRIPTION
**Pull request by kelebra**
This PR addresses issue #24519. Open questions so far:

- Does this one deserve on its own `class` level implementation (like `LazySource` and friends)?
- Should it be added to Cookbook (`stream-cookbook.md`)?

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-docs/src/main/paradox/stream/stages-overview.md [hunk 1]
```diff
@@ -201,6 +201,26 @@ Defers creation and materialization of a `Source` until there is demand.
...
+
+Defers creation and materialization of a `CompletionStage` until there is demand.
+
+**emits** the future completes
+
+**completes** after the future has completed
+
+---------------------------------------------------------------
+
+### actorPublisher
```
Comment (ktoso) [inline akka-docs/src/main/paradox/stream/stages-overview.md]: This should not be here anymore :)

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala [hunk 2]
```diff
@@ -250,6 +250,14 @@ object Source {
   def lazily[T, M](create: function.Creator[Source[T, M]]): Source[T, CompletionStage[M]] =
     scaladsl.Source.lazily[T, M](() ⇒ create.create().asScala).mapMaterializedValue(_.toJava).asJava
 
+  /**
+   * Creates a `Source` from supplied future factory that is not called until downstream demand. When source gets
+   * materialized the materialized future is completed with the value from the factory. If downstream cancels or fails
+   * without any demand the create factory is never called and the materialized `Future` is failed.
+   */
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Add a link to the normal one

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 3]
```diff
@@ -412,6 +412,14 @@ object Source {
   def lazily[T, M](create: () ⇒ Source[T, M]): Source[T, Future[M]] =
     Source.fromGraph(new LazySource[T, M](create))
 
+  /**
+   * Creates a `Source` from supplied future factory that is not called until downstream demand. When source gets
+   * materialized the materialized future is completed with the value from the factory. If downstream cancels or fails
+   * without any demand the create factory is never called and the materialized `Future` is failed.
+   */
+  def lazilyAsync[T](create: () ⇒ CompletionStage[T]): Source[T, Future[NotUsed]] =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: This should work on Scala.concurrent.Future :)

**Review by ktoso — COMMENTED** [review]
Almost there, please add a test too

File: akka-docs/src/main/paradox/stream/stages-overview.md (hunk 1 above)
Comment (kelebra) [reply akka-docs/src/main/paradox/stream/stages-overview.md]: Ooops, removed.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala (hunk 2 above)
Comment (kelebra) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Yep, done.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala (hunk 3 above)
Comment (kelebra) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: Makes sense, also done.

**Comment by kelebra** [comment]
Added unit tests similar to the ones from `LazySourceSpec`

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package akka.stream.scaladsl
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: All files must have the header:

```
/**
 * Copyright (C) 2014-2018 Lightbend Inc. <https://www.lightbend.com>
 */
```

We'll soon automate that I hope

Comment (kelebra) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: Added but also see that you did that already, sorry for kind of a mess :(

Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: Oh, sorry -- I somehow expected you be in EU timezone so was taking my timeprepping and going to merge this while EU is asleep :)

Comment (kelebra) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: CST actually :) but thanks for taking care of it, much appreciated

Comment (kelebra) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: Also, I could work on automating copyright check if you can supply issue for it

**Comment by ktoso** [comment]
>     Does this one deserve on its own class level implementation (like LazySource and friends)?

Nope, it's good as is.

>     Should it be added to Cookbook (stream-cookbook.md)?

Nope, it's enough to have it in the overview doc. Cookbook is for "this is how you can put together a number of things to achieve some thing"

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala (hunk 4 above)
Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazilyAsyncSpec.scala]: Cheers to CST from JST then :)

For the copyright we want to replicate what we have here: https://github.com/akka/akka-http/pull/1738

**Comment by ktoso** [comment]
LGTM, simple enough that we can merge once validated I think 👍

**Review by ktoso — APPROVED** [review]

**Comment by kelebra** [comment]
I think all issues highlighted in PR were addressed. Is there anything else required for this PR to be merged?

**Comment by ktoso** [comment]
Lookin good, thanks 👍

**Pull request merged**